vault/diagnose/tls_verification.go

```diff
@@ -214,3 +267,73 @@ func NearExpiration(c *x509.Certificate) (bool, time.Duration) {
 	}
 	return false, timeToExpiry
 }
+
+// TLSMutualExclusionCertCheck returns error if both TLSDisableClientCerts and TLSRequireAndVerifyClientCert are set
+func TLSMutualExclusionCertCheck(l *configutil.Listener) error {
+
+	if l.TLSDisableClientCerts {
+		if l.TLSRequireAndVerifyClientCert {
+			return fmt.Errorf("the tls_disable_client_certs and tls_require_and_verify_client_cert fields in the " +
+				"listener stanza of the vault server config are mutually exclusive fields. Please ensure they are not both set to true.")
+		}
+	}
+	return nil
+}
+
+// TLSClientCAFileCheck Checks the validity of a client CA file
+func TLSClientCAFileCheck(l *configutil.Listener) ([]string, error) {
+
+	if l.TLSDisableClientCerts {
+		return nil, nil
+	} else if !l.TLSRequireAndVerifyClientCert {
+		return nil, nil
+	}
+
+	var warningsSlc []string
+
+	// Parse TLS Certs from the tls config
+	leafCerts, interCerts, rootCerts, err := ParseTLSInformation(l.TLSClientCAFile)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(rootCerts) == 0 {
+		return nil, fmt.Errorf("No root cert found!")
+	}
+	if len(rootCerts) > 1 {
+		warningsSlc = append(warningsSlc, fmt.Sprintf("Found Multiple rootCerts instead of just one!"))
+	}
+
+	// Checking for Self-Signed cert and return an explicit error about it.
+	// Self-Signed certs are placed in the leafCerts slice when parsed.
+	if len(leafCerts) > 0 && !leafCerts[0].IsCA && bytes.Equal(leafCerts[0].RawIssuer, leafCerts[0].RawSubject) {
+		return warningsSlc, fmt.Errorf("Found a Self-Signed certificate!")
+	}
+
+	if len(interCerts) > 0 {
+		return nil, fmt.Errorf("Found at least one intermediate cert in a root CA cert.")
+	}
+
+	if len(leafCerts) > 0 {
+		return nil, fmt.Errorf("Found at least one leafCert in a root CA cert.")
+	}
+
+	// Adding rootCerts to leafCert to perform verification in TLSErrorChecks
+	leafCerts = append(leafCerts, rootCerts[0])
+
+	var warnings []string
+	// Check for TLS Warnings
+	warnings, err = TLSFileWarningChecks(leafCerts, interCerts, rootCerts)
+	warningsSlc = append(warningsSlc, warnings...)
+	if err != nil {
+		return warningsSlc, err
+	}
+
+	// Check for TLS Errors
+	if err = TLSErrorChecks(leafCerts, interCerts, rootCerts); err != nil {
```

Comment:
Could you finesse the error returned to indicate to replace all mention of leaves with root?